kyo-actor/shared/src/main/scala/kyo/Subject.scala

```diff
@@ -0,0 +1,169 @@
+      */
+    def trySend(message: A)(using Frame): Boolean < (IO & Abort[Closed])
+
+    /** Sends a message and waits for a response.
+      *
+      * This method implements the request-response pattern by automatically creating a temporary reply channel. It's useful when you need
+      * to get a response back after sending a message.
+      *
+      * For example, with a message type like:
+      *
+      * case class GetData(id: Int, replyTo: Subject[ResponseData])
+      *
+      * You can use:
+      *
+      * subject.ask(GetData(123, _))
+      *
+      * The returned type (ResponseData) is determined by the reply channel type in the message.
+      *
+      * @param f
+      *   A function that takes a reply Subject[B] and returns the message to send
+      * @tparam B
+      *   The type of response expected
+      * @return
+      *   The response of type B
+      */
+    def ask[B](f: Subject[B] => A)(using Frame): B < (Async & Abort[Closed]) =
+        for
+            promise <- Promise.init[Nothing, B]
+            _       <- send(f(Subject.init(promise)))
+            result  <- promise.get
+        yield result
+
+end Subject
+
+object Subject:
+
+    private val _noop =
+        init(
+            send = _ => (),
+            trySend = _ => false
+        )
+
+    /** Creates a no-operation Subject that discards all messages.
+      *
+      * This Subject implementation ignores any messages sent to it and performs no action. It can be useful for testing, as a placeholder,
+      * or when you need to explicitly discard messages.
+      *
+      * @tparam A
+      *   The type of messages this Subject can receive (and discard)
+      * @return
+      *   A Subject[A] that discards all messages
+      */
+    def noop[A]: Subject[A] = _noop.asInstanceOf[Subject[A]]
+
+    /** Creates a Subject that completes a Promise with each received message.
+      *
+      * This method creates a Subject that will attempt to complete the provided Promise with any message it receives. The Promise is
+      * completed with a successful Result containing the message.
+      *
+      * Important: Only the first message sent to this Subject will successfully complete the Promise. Any subsequent messages will result
+      * in an `Abort[Closed]` effect, indicating that the Subject is closed (since the Promise can only be completed once).
+      *
+      * @param promise
+      *   The Promise to complete when messages are received
+      * @tparam E
+      *   The error type of the Promise
+      * @tparam A
+      *   The type of messages this Subject can receive
+      * @return
+      *   A Subject[A] that completes the Promise with the first received message and aborts with Closed for subsequent messages
+      */
+    def init[E, A](promise: Promise[E, A])(using frame: Frame): Subject[A] =
+        def tryComplete(r: A) =
+            Abort.whenNot(promise.complete(Result.succeed(r)))(Closed("Subject", frame))
```

Comment:
I've updated `Abort.when` to allow effects in any of its parameters. Not allowing effects in the condition makes the API not very useful. I've also added `Abort.whenNot`.